flasher.cpp

```diff
@@ -255,8 +266,50 @@ void Flasher::LoopHandler()
             break;
         }
 
+        case FlasherStates::kLoadFirmwareFile:
+        {
+            // Load local firmware file
+            if (firmware_file_.size() != 0) {
+                file_content_ = firmware_file_.readAll();
+                firmware_file_.close();
+                firmware_file_.remove();
+                SetState(FlasherStates::kFlash);
+
+                // Load firmware file from url
+            } else {
+
+                DownloadFirmwareFromUrl();
+                emit ShowStatusMsg("Downloading");
+                timer_.start();
+                SetState(FlasherStates::kDownloadFirmwareFile);
+            }
+
+            break;
+        }
+
+        case FlasherStates::kDownloadFirmwareFile:
+        {
+            if (is_firmware_downloaded_) {
+                is_firmware_downloaded_ = false;
+                if (file_content_.isEmpty()) {
+                    emit ShowStatusMsg("Download error");
+                    SetState(FlasherStates::kIdle);
+                } else {
+                    SetState(FlasherStates::kFlash);
+                }
+            }
+
+            if (timer_.hasExpired(kTryToDownloadFirmwareTimeoutInMs)) {
```

Comment:
Please set this in else statement.
In theory, if downloading is done at the end of the timer expiration, state can be changed to kFlash and then it can be rewritten to kIdle.

Reply:
Very true!